runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -156,6 +154,172 @@ static TR::Node *lowerCASValues(
    }
 
 
+/**
+ * @brief Add checks to skip (fast-path) acmpHelper call
+ * 
+ * @details
+ * 
+ * This transformation adds checks for the cases where the acmp can be performed
+ * without calling the VM helper. The trasformed Trees represen the following operation:
+ *
+ * 1. If the address of lhs and rhs are the same, produce an eq (true) result
+ *    and skip the call (note the two objects must be the same regardless of
+ *    whether they are value types are reference types)
+ * 2. Otherwise, do VM helper call
+ *
+ * The transformation looks as follows:
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |treetop               |
+ *  |  icall acmpHelper    |
+ *  |    aload lhs         |
+ *  |    aload rhs         |
+ *  |ificmpeq --> ...      |
+ *  |  ==> icall           |
+ *  |  iconst 0            |
+ *  |BBEnd                 |
+ *  +----------------------+
+ *
+ *  ...becomes...
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |iRegStore x           |
+ *  |  iconst 1            |
+ *  |ifacmpeq  -->---------*---------+
+ *  |  aload lhs           |         |
+ *  |  aload rhs           |         |
+ *  |BBEnd                 |         |
+ *  +----------------------+         |
+ *  |BBStart (extension)   |         |
+ *  |iRegStore x           |         |
+ *  |  icall acmpHelper    |         |
+ *  |    aload lhs         |         |
+ *  |    aload rhs         |         |
+ *  |BBEnd                 |         |
+ *  +----------------------+         |
+ *        |                          |
+ *        +--------------------------+
+ *        |
+ *        v
+ *  +-----------------+
+ *  |BBStart
+ *  |ificmpeq --> ... |
+ *  |  iRegLoad x     |
+ *  |  iconst 0       |
+ *  |BBEnd            |
+ *  +-----------------+
+ *
+ */
+void
+J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt)
+   {
+   TR::Compilation* comp = self()->comp();
+   traceMsg(comp, "<LEO> Detected call to substitutabilityComparisonSymbol in n%dn\n", node->getGlobalIndex());
+
+   TR::CFG* cfg = comp->getFlowGraph();
+   cfg->setStructure(0);
+
+   // anchor call node after split point to ensure the returned value goes into
+   // either a temp or a global register
+   auto* anchoredCallTT = TR::TreeTop::create(comp, tt, TR::Node::create(TR::treetop, 1, node));
+   traceMsg(comp, "<LEO> Anchoring call node under treetop n%dn (0x%p)\n", anchoredCallTT->getNode()->getGlobalIndex(), anchoredCallTT->getNode());
+
+   // anchor the call arguments just before the call
+   // this ensures the values are live before the call so that we can
+   // propagate their values in global registers if needed
+   auto* anchoredCallArg1TT = TR::TreeTop::create(comp, tt->getPrevTreeTop(), TR::Node::create(TR::treetop, 1, node->getFirstChild()));
+   auto* anchoredCallArg2TT = TR::TreeTop::create(comp, tt->getPrevTreeTop(), TR::Node::create(TR::treetop, 1, node->getSecondChild()));
```

Comment:
This looks like https://github.com/eclipse/omr/blob/87bb94e10fc847bc89d1a55b866d6952e31bcea8/compiler/optimizer/OMROptimization.cpp#L214

Perhaps this pass would make more sense as a late MustBeDone optimization pass rather than in the CG itself since it is going to be doing some serious surgery that the optimizer might be able to help with? Just a thought...

Reply:
Not sure. When I discussed this with others, the suggestion was to do it in lower trees.

Reply:
Thinking on it a little more, it would make it easier to work around the strange interaction between regDepCopyRemoval and the post GRA block splitter.

Then again, this code is really more "lowering" than "optimizing", so it makes sense for it to be in lower trees.